File: SRM684Div2/600/solution.cpp
#include <vector>
#include <queue>
using namespace std;
typedef long long ll;

class DivFreed2 {
private:
    const ll P = 1000000007;
public:
    int count (int n, int k) {
        vector<vector<ll>> dp(n + 1, vector<ll>(k + 1, 0));
        for (int i = 1; i <= k; i++) {
            dp[1][i] = 1;
        }
        
        for (int i = 2; i <= n; i++) {
            ll sum = 0;
            for (int j = 1; j <= k; j++) {
                sum += dp[i-1][j];
                sum %= P;
            }
            
            for (int j = 1; j <= k; j++) {
                dp[i][j] = sum;
                for (int m = 2 * j; m <= k; m += j) {
                    dp[i][j] = (dp[i][j] + P - dp[i - 1][m]) % P;
                }
            }
        }
        
        ll res = 0;
        for (int i = 1; i <= k; i++) {
            res += dp[n][i];
            res %= P;
        }
        return (int)res;
    }
};

File: SRM684Div2/600/solution.py
class DivFreed2():
    def count(self, n, k):
        mod = 1000000007

        ans = [1] * (k + 1)
        ans[0] = 0
        for i in range(n - 1):
            next = [0] * (k + 1)
            s = sum(ans)
            for j in range(1, k + 1):
                next[j] = s
                for m in range(j + j, k + 1, j):
                    next[j] -= ans[m]
                next[j] %= mod
            ans = next
        return sum(ans) % mod
